py,esp32,rp2: Add new cstack API, check margin macro, fix esp32 stack check.

projectgus opened the pull request:
### Summary

* Motivation for this was testing the ESP32-C3 bug fix in https://github.com/micropython/micropython/pull/15583/ and realising there is a subtle race condition on esp32 port where the stack check limit may not be set reliably, depending on function timing.
* On looking at the whole situation, every port adds some "stack limit margin" but it's usually hard-coded in the call to `mp_stack_set_limit()`. Threaded ports have the extra challenge of adding a margin in the thread module. However only the value passed to `mp_stack_set_limit()` needs to have the margin applied (i.e. if the user sets a thread stack size of N bytes, the tasks will have a real stack size of N bytes and the stack limit checker sets its value to `N - margin` bytes.)
* After discussing with Damien this PR has expanded to add both a `MICROPY_STACK_CHECK_MARGIN` macro and a new cstack API which sets the stack origin and the stack limit (based off the stack size) in one call.
* To offset the code size impact on ports without threads, new cstack API can be inlined in some cases.
* Old stackctrl.h API is kept unchanged, apart from a note about deprecation and macros to disable when building `MICROPY_PREVIEW_VERSION_2`.
* `esp32`, `unix` and `rp2` ports are converted in this PR as initial examples.

### Testing

* Ran `stress` and `thread` tests on unix port, esp32 port (S3 and C3), and on rp2 port.
* A commit in this PR enables the `stress` tests for the esp32 port.

### Trade-offs and Alternatives

Could have made a much smaller fix for the esp32 port race condition by adjusting the stack size before creating the thread, or by applying a mutex. However this approach seems easier to understand overall, and implements a single source of truth for the stack limit margin.
 
### Follow-up work

* Refactor other ports to use the new API.
* Suspect the stack limitmargin is set quite conservatively (especially for ESP32-C3 where it is 2KB). Could experiment with reducing this to save some RAM.

projectgus committed:
tests/run-tests.py: Enable stress tests on esp32 port.

Now passing on ESP32-S3 and ESP32-C3.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
py: Add new cstack API for stack checking, with limit margin macro.

Currently the stack limit margin is hard-coded in each port's call to
`mp_stack_set_limit()`, but on threaded ports it's fiddlier and can lead to
bugs (such as incorrect thread stack margin on esp32).

This commit provides a new API to initialise the C Stack in one function
call, with a config macro to set the margin.  Where possible the new call
is inlined to reduce code size in thread-free ports.

Intended replacement for `MP_TASK_STACK_LIMIT_MARGIN` on esp32.

The previous `stackctrl.h` API is still present and unmodified apart from a
deprecation comment.  However it's not available when the
`MICROPY_PREVIEW_VERSION_2` macro is set.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
rp2: Switch to use new cstack API for stack limit checks.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
esp32: Fix thread stack limit margin, change to new cstack API.

This change moves that complexity out into the stack checker and fixes the
bug where stack margin wasn't set correctly by ESP32-C3 threads.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

dpgeorge commented on py/stackctrl.h:
```diff
@@ -26,15 +26,20 @@
 #if MICROPY_STACK_CHECK
 
-void mp_stack_set_limit(mp_uint_t limit);
+inline static void mp_stack_set_limit(mp_uint_t limit) {
```
This is no longer setting the limit, instead it's setting the total available stack space.  The limit is then that number minus the margin.

So I suggest renaming this function to something like `mp_stack_set_size()`.  If we do that, it would force users to notice this change in API.  OTOH maybe it's nicer for them to provide a backwards compatible macro `#define mp_stack_set_limit mp_stack_set_size`?

projectgus replied:
Good point... what about `mp_stack_check_init(mp_uint_t stack_size)`. Then the name still indicates the purpose, but more accurate about what it will do.

dpgeorge replied:
I think `mp_stack_check_init()` is confusing because there's already `mp_stack_ctrl_init()`... with two init functions it's confusing which one to call.

The current situation is that you either do one of the following to initialise stack checking:
- call `mp_stack_ctrl_init()` to use the current stack pointer to set the top of the stack, then call `mp_stack_set_limit()` to set the limit from there
- call `mp_stack_set_top()` to set an explicit top pointer, then call `mp_stack_set_limit()`

Then you can call `mp_stack_usage()` and `mp_stack_check()`.

If we are going to rejig the API, I'd propose to have two separate init functions:
- `mp_stack_check_init_with_sp_here(size_t stack_size)`
- `mp_stack_check_init_with_top(void *top, size_t stack_size)`

(Note the `size_t` type.)

Then it's much clearer what to call, and you only need to call one function to init it.

dpgeorge commented on py/stackctrl.h:
```diff
@@ -26,15 +26,20 @@
 
-void mp_stack_set_limit(mp_uint_t limit);
+inline static void mp_stack_set_limit(mp_uint_t limit) {
+    assert(limit > MICROPY_STACK_CHECK_MARGIN); // Should be enforced by port
```
Suggest using `MP_STATIC_ASSERT` (if possible here...).

projectgus replied:
Not possible here, limit is not guaranteed to be compile-time constant.

However the compiler will be smart, if assertions are enabled and limit *is* constant then the assert will be elided if it passes, and if it fails then dead code elimination will conveniently remove everything that comes after. :laughing:

dpgeorge replied:
Aah, right.

dpgeorge commented:
An alternative to this would be to still have the `MICROPY_STACK_CHECK_MARGIN` macro but use it explicitly at the call site for `mp_stack_set_limit()`.  Eg in `mp_thread_init_state()` it would do:
```c
    mp_stack_set_limit(stack_size - MICROPY_STACK_CHECK_MARGIN);
```
That makes it more obvious that the limit is the stack size minus an explicit margin.  Otherwise the margin is kind of hidden in the `stackctrl.h` implementation.

But, this alternative approach is more verbose (more duplicated code with the subtraction of the margin everywhere) so probably not any better than what this PR does.

dpgeorge commented on py/stackctrl.h:
```diff
@@ -26,15 +26,20 @@
 #define MICROPY_INCLUDED_PY_STACKCTRL_H
 
-#include "py/mpconfig.h"
+#include "py/mpstate.h"
```
I was a bit worried this would introduce a circular header dependency, but luckily it does not.

So, now it would also be possible to make `mp_stack_set_top()` static-inline, which should also help to reduce code size.

projectgus replied:
Have pushed a commit here to do this as well.

projectgus committed:
unix: Switch stack limit check to new cstack API.

Necessary to pass CI when testing the V2 preview APIs.

Also adds an extra coverage test for the legacy stackctrl API, to maintain
coverage and check for any regression.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge Have updated according to our discussion earlier, and updated the PR description.

The project coverage check is failing because stackctrl.c is no longer covered. What's the best way to deal with that?

dpgeorge commented:
> The project coverage check is failing because stackctrl.c is no longer covered. What's the best way to deal with that?

Hmmmm... three options I can think of:
1. Add code to `ports/unix/coverage.c` to call the old functions to get coverage.
2. Add a `#if MICROPY_STACK_CHECK_LEGACY` around everything in `py/stackctrl.c` and disable this on unix coverage builds, to exclude that code from coverage checks.
3. Implement the new functions in terms of the old ones.

Option 3 makes it harder to easily remove the old functions when the time comes.  Probably option 1 is the best.

projectgus commented:
> Option 3 makes it harder to easily remove the old functions when the time comes. Probably option 1 is the best.

Thanks, have push a commit to try that approach (and coverage seems happy, at least.)

dpgeorge commented on ports/rp2/mpconfigport.h:
```diff
@@ -158,6 +158,8 @@
 // Number 3 is currently used by pico-sdk (PICO_TIME_DEFAULT_ALARM_POOL_HARDWARE_ALARM_NUM)
 #define MICROPY_HW_SOFT_TIMER_ALARM_NUM         (2)
 
+#define MICROPY_STACK_CHECK_MARGIN              (256)
```
Please move this config option to where it is relative to how it's ordered in `py/mpconfig.h`.

That would mean placing it between `MICROPY_ENABLE_GC` and `MICROPY_ENABLE_EMERGENCY_EXCEPTION_BUF`.

projectgus replied:
Done. Do you want the same thing for esp32 mpconfigport.h? It would mean splitting up a block of defines with an ifdef.

dpgeorge replied:
Yes, please move it.  Consistency is better here.

projectgus replied:
Done!

dpgeorge commented on py/cstack.c:
```diff
@@ -0,0 +1,56 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014 Paul Sokolovsky, 2024 Angus Gratton
```
The convention in this project is to put separate authors' copyright on separate lines, ie:
```c
* Copyright (c) 2014 Paul Sokolovsky
* Copyright (c) 2024 Angus Gratton
```

dpgeorge commented on py/cstack.h:
```diff
@@ -0,0 +1,62 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014 Paul Sokolovsky, 2024 Angus Gratton
```
separate lines, as above

dpgeorge commented on py/cstack.h:
```diff
@@ -0,0 +1,62 @@
+
+#include "py/mpstate.h"
+
+// Both init functions accept the full stack size. Set the
```
"Both init functions below accept..."

dpgeorge commented on py/stackctrl.c:
```diff
@@ -25,6 +25,11 @@
 
 #include "py/runtime.h"
+
+// This API is deprecated, please use py/cstack.h instead
```
Suggest putting this comment above the include of py/runtime.h, to mirror `stackctrl.h`.

dpgeorge commented:
This is looking really good now!  Just a few minor comments.

dpgeorge commented:
Thanks for updating, now merged.

dpgeorge commented:
@projectgus looks like this may have broken `multi_bluetooth/ble_irq_calls.py` on ESP32-C3.  That test tests for recursion within a BLE IRQ handler, and now it seems that ESP32-C3 has less stack for Python calls because of the increased margin, so the test fails.

The BLE IRQ handler on esp32 has `CONFIG_BT_NIMBLE_TASK_STACK_SIZE=6144` minus 2048 minus `MICROPY_STACK_CHECK_MARGIN` bytes.  So that's very small on ESP32-C3/RISC-V.

To fix it I suggest decreasing `MICROPY_STACK_CHECK_MARGIN` on RISC-V to a less conservative value.  If that still doesn't fix `multi_bluetooth/ble_irq_calls.py` then we should probably increase `CONFIG_BT_NIMBLE_TASK_STACK_SIZE` on RISC-V.
